dll/win32/shell32/dialogs/filetypes.cpp

```diff
@@ -1614,56 +1856,56 @@ FileTypesDlg_OnDelete(HWND hwndDlg)
         FileTypesDlg_RemoveExt(hwndDlg);
 
         // Select first item (Win2k3 does it)
-        LV_ITEMW item = { LVIF_STATE };
-        item.stateMask = item.state = LVIS_FOCUSED | LVIS_SELECTED;
-        item.iItem = 0;
-        ListView_SetItem(GetDlgItem(hwndDlg, IDC_FILETYPES_LISTVIEW), &item);
+        HWND hListView = GetDlgItem(hwndDlg, IDC_FILETYPES_LISTVIEW);
+        ListView_SetItemState(hListView, 0, -1, LVIS_FOCUSED | LVIS_SELECTED);
     }
 }
 
 static void
-FileTypesDlg_OnItemChanging(HWND hwndDlg, PFILE_TYPE_ENTRY pEntry)
+FileTypesDlg_OnItemChanging(HWND hwndDlg, PFILE_TYPE_ENTRY pEntry, PFILE_TYPE_GLOBALS g)
 {
-    WCHAR Buffer[255];
-    static HBITMAP s_hbmProgram = NULL;
-
-    // format buffer and set groupbox text
-    CStringW strFormat(MAKEINTRESOURCEW(IDS_FILE_DETAILS));
-    StringCbPrintfW(Buffer, sizeof(Buffer), strFormat, &pEntry->FileExtension[1]);
-    SetDlgItemTextW(hwndDlg, IDC_FILETYPES_DETAILS_GROUPBOX, Buffer);
+    HBITMAP &hbmProgram = g->hOpenWithImage;
+    LPCWSTR DispAssoc = pEntry->GetAssocForDisplay();
+    LPCWSTR TypeName = GetTypeName(pEntry, g);
+    CStringW buf;
 
     // format buffer and set description
-    strFormat.LoadString(IDS_FILE_DETAILSADV);
-    StringCbPrintfW(Buffer, sizeof(Buffer), strFormat,
-                    &pEntry->FileExtension[1], pEntry->FileDescription,
-                    pEntry->FileDescription);
-    SetDlgItemTextW(hwndDlg, IDC_FILETYPES_DESCRIPTION, Buffer);
+    FileTypesDlg_SetGroupboxText(hwndDlg, DispAssoc);
+    if (pEntry->IsExtension())
+        buf.Format(IDS_FILE_DETAILSADV, DispAssoc, TypeName, TypeName);
+    else
+        buf = L"";
+    SetDlgItemTextW(hwndDlg, IDC_FILETYPES_DESCRIPTION, buf.GetString());
 
     // delete previous program image
-    if (s_hbmProgram)
+    if (hbmProgram)
     {
-        DeleteObject(s_hbmProgram);
-        s_hbmProgram = NULL;
+        DeleteObject(hbmProgram);
+        hbmProgram = NULL;
     }
 
+    // set program name
+    LPCWSTR appname = GetAppName(pEntry);
+    SetDlgItemTextW(hwndDlg, IDC_FILETYPES_APPNAME, appname);
+
     // set program image
     HICON hIconSm = NULL;
-    ExtractIconExW(pEntry->ProgramPath, 0, NULL, &hIconSm, 1);
-    s_hbmProgram = BitmapFromIcon(hIconSm, 16, 16);
+    LPCWSTR exe = GetProgramPath(pEntry);
+    if (*exe)
+    {
+        ExtractIconExW(exe, 0, NULL, &hIconSm, 1);
+    }
+    hbmProgram = BitmapFromIcon(hIconSm, 16, 16);
```

Comment:
Shouldn't this "16" be `GetSystemMetrics(SM_CXSMICON)` and `GetSystemMetrics(SM_CYSMICON)`  ?

Reply:
Existing code.

This code should not exist at all, why the hell is it converting to a bitmap when the static control supports icons?

The bitmap also looks wrong when themes are used.

The PR is already large enough, not fixing this bitmap thing now.

Reply:
OK then no need for change.